feat(to-react-native): To React Native parser

robrechtme opened the pull request:
## What it does

Hi 👋 

Here are the first steps to a `to-react-native` parser. It is based on the `to-jss` parser, but has a few differences since CSS (JSS) and React Native styles aren't fully compatible. 


## Additional information

🚧, This PR is still in WIP, though I would love to hear your thoughts/feedback. 

**TO DO**: 
- [x] Write tests 
- [x] Improve `README`


## Linear ticket
Resolves [DEV-XXX]

ChucKN0risK commented on parsers/to-react-native/README.md:
```diff
@@ -0,0 +1,150 @@
+| `colorFormat`      | optional | `rgb`, `prgb`, `hex`, `hex6`, `hex3`, `hex4`, `hex8`, `name`, `hsl`, `hsv` | `rgb`       | The format of all colors (gradients, borders, textStyles etc.)                                                        |
+| `objectName`       | optional | `string`                                                                   | `'theme'`   | The name of the JS object containing the theme,                                                                       |
+| `assetsFolderPath` | optional | `string`                                                                   | `undefined` | The relative location of the folder to import the assets from, if not provided, the assets will be referenced by URL. |
+| `prettierConfig`   | optional | [`prettier.Config`](https://prettier.io/docs/en/options.html)              | `undefined` | Formatting options of the file.                                                                                       |
```
Please list all the Prettier config properties this parser supports. You can find an example of how we display this in the `options` section here: https://github.com/Specifyapp/parsers/blob/master/parsers/to-scss-mixin-text-style/README.md#interface

Thanks 🙏

ChucKN0risK commented on parsers/to-react-native/README.md:
```diff
@@ -0,0 +1,150 @@
+| `objectName`       | optional | `string`                                                                   | `'theme'`   | The name of the JS object containing the theme,                                                                       |
+| `assetsFolderPath` | optional | `string`                                                                   | `undefined` | The relative location of the folder to import the assets from, if not provided, the assets will be referenced by URL. |
+| `prettierConfig`   | optional | [`prettier.Config`](https://prettier.io/docs/en/options.html)              | `undefined` | Formatting options of the file.                                                                                       |
+
```
Please set `## Types` as the third section of the README (i.e. https://github.com/Specifyapp/parsers/blob/master/parsers/to-scss-mixin-text-style/README.md#types).

This section must include the types of data that is accepted as input by the parser and the type of data that is returned by the parser. This helps our user understand which parser can be chained with each other.

End result 👇

---

## Types

ℹ️ **Please be aware that, depending on the order you use parsers, their input and output types have to match.**

### Input

Array of object with at least name, value and type

```ts
Array<{name: string, value: any, type: string}>
```

### Output

```ts
string
```

ChucKN0risK commented on parsers/to-react-native/README.md:
```diff
@@ -0,0 +1,150 @@
+| `assetsFolderPath` | optional | `string`                                                                   | `undefined` | The relative location of the folder to import the assets from, if not provided, the assets will be referenced by URL. |
+| `prettierConfig`   | optional | [`prettier.Config`](https://prettier.io/docs/en/options.html)              | `undefined` | Formatting options of the file.                                                                                       |
+
+### Before/After
```
Please include the example of usage inside a "Usage" section.

Also, please provide the config passed to the parser to return the mentioned output. I'vbe created the section for you but you need to update it to provide the options you used.

Here's the end result:

---

## Usage

### Config
```json
{
  "name": "to-jss",
  "options": {
  
  }
}
```

### Before/After

#### Input

```js
[
  {
    name: 'activity.svg',
    value: {
      url: 'https://s3-us-west-2.amazonaws.com/figma-alpha-api/img/a114/ce5c/947dcb83ea93c2da18ee2ea16f470a30',
    },
    type: 'vector',
  },
  {
    name: 'Body',
    value: {
      font: {
        meta: {
          source: 'localStyles',
        },
        name: 'Inter-Medium',
        type: 'font',
        value: {
          isItalic: false,
          fontFamily: 'Inter',
          fontWeight: 500,
          fontPostScriptName: 'Inter-Medium',
        },
      },
      color: {
        value: {
          a: 1,
          b: 196,
          g: 196,
          r: 196,
        },
      },
      fontSize: {
        value: {
          unit: 'px',
          measure: 14,
        },
      },
      textAlign: {
        vertical: 'top',
        horizontal: 'left',
      },
      lineHeight: {
        value: {
          unit: 'px',
          measure: 20,
        },
      },
    },
    type: 'textStyle',
  },
  {
    name: 'Colors / Accent',
    value: {
      a: 1,
      b: 239,
      g: 80,
      r: 102,
    },
    type: 'color',
  },
];
```

#### Output

```js
import acmeLogo from '../src/assets/acmeLogo.png';
import activity from '../src/assets/activity.svg';

const theme = {
  bitmap: {
    acmeLogo: acmeLogo,
  },
  vector: {
    activity: activity,
  },
  depth: {
    background: {
      elevation: 1,
      shadowOpacity: 0.1815,
      shadowRadius: 0.54,
      shadowOffset: { width: 0.6, height: 0.6 },
    },
  },
  duration: {
    base: 300,
  },
  measurement: {
    baseSpace01: 4,
  },
  textStyle: {
    body: {
      fontWeight: 500,
      fontSize: 14,
      lineHeight: 20,
      fontFamily: 'Inter-Medium',
      color: '#1e212b',
      letterSpacing: 10,
    },
  },
  border: {
    borderAccent: {
      borderWidth: 2,
      borderStyle: 'solid',
      borderColor: 'rgb(102, 80, 239)',
      borderRadius: 16,
    },
  },
  color: {
    colorsAccent: 'rgb(87, 124, 254)',
  },
  font: {
    firaCodeMedium: 'FiraCode-Medium',
  },
  gradient: {
    gradientsColored: [
      {
        angle: 90,
        colors: ['rgb(245, 72, 63)', 'rgb(255, 142, 5)'],
        locations: [0, 1],
      },
    ],
  },
  opacity: { subtle: 0.5, transparent: 0.1, visible: 0.95 },
};
export default theme;
```

ChucKN0risK commented:
Thanks a lot for your contribution @robrechtme 🙏

I commented some parts of the README so you can easily update it.

robrechtme commented:
@ChucKN0risK I updated the `README` (sorry for ignoring the [How to create your own parser?](https://github.com/Specifyapp/parsers#how-to-create-your-own-parser) ) and added a basic snapshot test. 
I would love to add some "proper" tests when I have time.

AMoreaux commented on parsers/to-react-native/tokens/gradient.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  toReactNative({ colorFormat = 'rgb' }: OptionsType = {}) {
+    const value: any[] = this.value.gradients.map(gradient => {
```
Prefer the usage of typescript inference instead of `any` type.
You can remove `any[]` safely.

robrechtme replied:
Whoops that's lazy on my behalf 🙈

AMoreaux reviewed (changes requested):
Thanks a lot for your PR 😊

To be able to merge and deploy the parser, we need several small updates:
- Squash your PR to have only one commit. (To let us keeping the git history as clean as possible)
- Add more tests on `vector`, `bitmap`, and `duration`. see:
![image](https://user-images.githubusercontent.com/7237157/136044961-239b4a99-9462-4334-b9ad-caca5234412e.png)

AMoreaux commented:
Currently, GitHub actions do not work properly (https://www.githubstatus.com/) so tests cannot be executed at this moment.

robrechtme commented on parsers/to-react-native/to-react-native.parser.ts:
```diff
@@ -0,0 +1,95 @@
+    const template = new Template(pattern);
+    let imports = '';
+    const styles = Object.keys(tokensGroupByType)
+      .filter(type => type !== 'shadow') // Only support depth for now
```
It will create an empty `shadows: {}` entry in the output, which is undesired I think.

AMoreaux replied:
You're right. 
I suggest omitting all types that include empty content instead of filter only on `shadow`. 
To do that you could remove the filter and wrap the concatenation of the result in a condition like:
```ts
if (content) {
  result += `${_.camelCase(type)}: {${content}},`;
}
```

AMoreaux commented on parsers/to-react-native/README.md:
````diff
@@ -0,0 +1,350 @@
+
+```ts
+interface parser {
+  name: 'to-jss';
````
I missed that. Please change `to-jss` in `to-react-native`

AMoreaux reviewed (changes requested):
Last update 😇 

PS: I notice that the current seeds have not let you tests durations properly. We will work on it.

AMoreaux commented:
Thanks, @robrechtme I will deploy the parser on staging, test it and if everything is ok, deploy it on production.
I will tell you when it will be available.

robrechtme commented:
Thanks a lot for the quick & clear response 🚀 Can't wait to use it in our projects.
